### PR TITLE
WIP - Large custom resource benchmark

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/pull/102898 fixed a performance regression for large custom resources, but even with the regression fixed, we want to further improve the performance of large custom resources.

This adds benchmarks for large custom resources both with and without schemas.

I was quite surprised to see that the initial results seem to suggest that custom resources with schemas perform much better than those without:

```
go test ./vendor/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager -run=NONE -bench=BenchmarkUpdateLargeCustomResource -memprofile=mem.prof -cpuprofile=cpu.prof -benchtime=30s -v
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager
BenchmarkUpdateLargeCustomResource
BenchmarkUpdateLargeCustomResource/example.k8s.io/v1,_Kind=NoSchema
BenchmarkUpdateLargeCustomResource/example.k8s.io/v1,_Kind=NoSchema-4         	     100	 349087704 ns/op	66563334 B/op	 1373774 allocs/op
BenchmarkUpdateLargeCustomResource/example.k8s.io/v1,_Kind=LargeStatus
BenchmarkUpdateLargeCustomResource/example.k8s.io/v1,_Kind=LargeStatus-4      	     656	  54366761 ns/op	19971863 B/op	  305308 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager	79.142s

```

TODO:
- There are significant differences between these numbers and what I've observed with a [manual test](https://gist.github.com/jpbetz/c06ee61492d70e47d9455af5f85dd9de) against a actual kubernetes cluster for similar custom resources. The manual test shows 8.5 s latency vs 0.35s latency seen here for a similar custom resource. I would like to get this benchmark to better simulate that observed e2e latency...
- Figure out why CRDs without schemas are an order of magnitude slower than custom resources with schemas in this benchmark.
- Benchmark large CRD schemas, which are sometimes avoided for performance reasons (e.g. https://github.com/argoproj/argo-workflows/blob/master/manifests/base/crds/full/README.md).

```release-note
NONE
```

### CODE REVIEW

**Comment by jpbetz** [comment]
@lavalamp @apelisse the numbers here are a bit surprising to me. It's quite possible there is something not quite right about the benchmark itself. Do you see anything wrong/missing?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        },
+        "status": {
+          "type": "object",
+          "properties": {
+            "items": {
+              "type": "object",
+              "x-kubernetes-map-type": "granular",
+              "additionalProperties": {
+                "type": "object",
+                "properties": {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:43]: is this a large status schema?

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:43]: yeah, I don't know if it matters if it is spec of status for this benchmark. The reason I ended up with status here is that CRD I was modeling this off of  (argo workflow) had a large status.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:43]: I wouldn't be opposed to moving this to the spec.

The update call used here is a low level SSA one that can modify either spec of status.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json [hunk 2]
```diff
@@ -0,0 +1,252 @@
+{
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:1]: file comment: put this in a testdata/ directory?

**Comment by lavalamp** [comment]
I agree these numbers are surprising -- I'd e.g. put a panic in the innermost field manager to get a stack trace, and then compare the stack trace from a real cluster and this test to see if we have missed some part of the field manager stack?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json (hunk 1 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:43]: Oh, @liggitt were you asking why the schema isn't that large? I've got this PR mis-phrased, I'm trying to demonstrate a large custom resource, the CRD is pretty small. I'll update a few things to clear that up.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/crd-schemas.json:43]: yes, that was what I was asking... got it

**Comment by jpbetz** [comment]
> I agree these numbers are surprising -- I'd e.g. put a panic in the innermost field manager to get a stack trace, and then compare the stack trace from a real cluster and this test to see if we have missed some part of the field manager stack?

Good idea. I tried this. Surprisingly, they were almost the same. Only difference was that `NewCapManagersManager` was missing from the test stack, so I added it, but none of the existing tests hit the cap so it doesn't change anything.

The benchmark seems to suggest that validating schemas using the "derived" SMD schema (which is the "allow everything" schema we use when either a CRD does not have a schema, or when `x-kubernetes-preserve-unknown-fields` is used for a branch of the schema) is far more expensive than validating with a fully defined schema.

We can't necessarily ask developers fully define their schemas would fix the latency problems though. Sometimes when the schema is fully defined, it is enormous, which causes other problem problems. (This was true for the use cases that instigated this investigation).

I suspect that validating using "derived" schemas could be optimized. In fact, I'm a bit unclear on what the validation pass achieves, since, as far as I can tell, all data conforms to the derived schema. So it might be possible to skip it entirely?

(I'm aware that toward our goal of promoting the use of schemas in CRDs, this doesn't help-- it makes the use of "schemaless" CRDs even more attractive for CRDs with large schemas and/or large CRs).

**Comment by apelisse** [comment]
> There are significant differences between these numbers and what I've observed with a manual test against a actual kubernetes cluster for similar custom resources.

We've spent a year suffering from this problem. In general, my understanding is that it's hard to isolate "memory allocation latency problems" in unit-test, while it's very frequent in an e2e test. We've never really managed to reproduce locally the problems we had seen in scale-tests or in e2e tests.

I can help you look at what's going on with the "deduced" type to see what's causing so many allocations.

**Comment by jpbetz** [comment]
Closing this to reflect that I am not actively working on it

**Pull request closed without merging**